### PR TITLE
[fix] Conditionally import CrossEncoder based on availability

### PR DESCRIPTION
**Pull request by bhavnicksm**
## TL;DR

Fixing the error where `CrossEncoder` was imported from `sentence_transformers` by defualt, even in cases when `korok[rerankers]` was not installed, causing `import korok` to error out.

image

## Autogen Summary
This pull request includes changes to the `CrossEncoderReranker` class in the `korok/rerankers/rerankers.py` file to improve error handling and modularity. The most important changes include adding a check for the availability of the `sentence_transformers` package and raising an appropriate error if it is not installed.

Improvements to error handling and modularity:

* Added an import statement for `importlib` to check for the availability of the `sentence_transformers` package.
* Added a docstring to the `CrossEncoderReranker` class to provide a brief description.
* Implemented a check in the `__init__` method to raise an `ImportError` if `sentence_transformers` is not installed, with instructions for installation.
* Added the `is_available` class method to check if the `sentence_transformers` package is available.

### CODE REVIEW

File: korok/rerankers/rerankers.py [hunk 1]
```diff
@@ -1,18 +1,30 @@
...
+    """A reranker class that uses a cross-encoder to rerank the results."""
+
     def __init__(self, model_name: str = "cross-encoder/ms-marco-MiniLM-L-6-v2", max_length: int = 512) -> None:
         """
         Initialize the CrossEncoder reranker.
 
         :param model_name: The name of the pre-trained cross-encoder model.
         :param max_length: The maximum sequence length for the cross-encoder.
         """
+        if not self.is_available():
```
Comment (stephantul) [inline korok/rerankers/rerankers.py:15]: IMO, checking whether an import is available is not a responsibility of the class. If you would like to tie requirements directly to classes, I would add a list of requirements to the class, which can then be imported/checked by a utility function.

File: korok/rerankers/rerankers.py [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
+            raise ImportError("sentence_transformers is not installed. Please install it using `pip install korok[rerankers]`.")
+        else:
+            global CrossEncoder
```
Comment (stephantul) [inline korok/rerankers/rerankers.py:18]: I'm not sure if the global is necessary here. Do we need it outside of the scope of this function?

**Review by stephantul — COMMENTED** [review]
Hey Bhavnick, thanks a lot for this PR. Overall, I think we would like to shy away from putting imports in classes (unless we really care about speed), as this ties the dependency structure of the package to class implementations.

In model2vec, we created a utility function for this. At the top of each file, we just attempt to import everything we need, and refuse to go on unless everything is importable. To do that here, we would need a conditional import of the reranker submodule, which is another architectural issue. 

Overall, I would suggest this requires a more extensive rewrite of the way we integrate components. I would personally lean towards not having the reranker be an attribute of a korok, but to create a subclass that has a reranker. An alternative could be to just make `sentence-transformers` a standard dependency. 

I'll wait for what @Pringled has to say, but I'd be in favor of not having conditional imports for now, and using `sentence-transformers` by default. We can refactor that away into an optional dependency once we settle on a specific architecture.

File: korok/rerankers/rerankers.py (hunk 2 above)
Comment (bhavnicksm) [reply korok/rerankers/rerankers.py:18]: Well, you can also pass it around inside the class via `self.CrossEncoder`, but usually imports inside functions are limited to the function scope.

**Comment by Pringled** [comment]
I agree with @stephantul, I think we can just default to sentence-transformers for now and avoid conditional imports. Since we are still thinking a bit about how to shape this package, that's easier for now, and we can always change that later.

**Comment by bhavnicksm** [comment]
Sure @Pringled and @stephantul! 

Thanks for the detailed explanations! 😊

Makes sense, I'll close this PR in that case. 

Thanks! ☺️
